### PR TITLE
benchmark: add support for payload size curves

### PR DESCRIPTION
**Pull request by adtac**
Closes https://github.com/grpc/grpc-go/issues/3236

### CODE REVIEW

**Comment by adtac** [comment]
Example output:

```bash
unary-networkMode_none-bufConn_false-keepalive_false-benchTime_5s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_1-payloadCurveFile_b508b89f69ff9233de912823b5a24fb3a4ad7a3fadba78cc6cfaa6f19dadc348-compressor_off-channelz_false-preloader_false:
50_Latency: 120.7220µs	90_Latency: 178.8390µs	99_Latency: 411.0110µs	Avg_Latency: 131.5520µs	Bytes/op: 15583.83175449244	Allocs/op: 222.66501306171992	
Histogram (unit: µs)
Count: 37897  Min:  61.8  Max: 6702.1  Avg: 131.55
------------------------------------------------------------
[     61.847000,      61.848000)      1    0.0%    0.0%  
[     61.848000,      61.852728)      0    0.0%    0.0%  
[     61.852728,      61.879813)      0    0.0%    0.0%  
[     61.879813,      62.034959)      1    0.0%    0.0%  
[     62.034959,      62.923671)     44    0.1%    0.1%  
[     62.923671,      68.014427)   2112    5.6%    5.7%  #
[     68.014427,      97.175466)   8076   21.3%   27.0%  ##
[     97.175466,     264.216744)  26940   71.1%   98.1%  #######
[    264.216744,    1221.068373)    696    1.8%   99.9%  
[   1221.068373,    6702.139000)     26    0.1%  100.0%  
[   6702.139000,            inf)      1    0.0%  100.0%  
Number of requests:  37897	Request throughput:  0 bit/s
Number of responses: 37897	Response throughput: 0 bit/s

```

File: benchmark/stats/curve.go [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+package stats
+
+import (
+	"crypto/sha256"
+	"encoding/binary"
+	"encoding/csv"
+	"encoding/hex"
+	"fmt"
+	"math"
+	"math/rand"
```
Comment (dfawley) [inline benchmark/stats/curve.go]: Please use internal/grpcrand instead.

Comment (adtac) [reply benchmark/stats/curve.go]: How are grpcrand's concurrency semantics useful here? Or are we interested in using grpcrand's random seed for testing reasons? Using a global lock seems wasteful on the surface, so I'm just trying to understand why it'd be useful here.

Comment (dfawley) [reply benchmark/stats/curve.go]: It's enforced by `vet.sh` that `math/rand` may not be imported (that's why travis is red).  I'd rather update this to use `grpcrand` than add an exception in vet.

Comment (dfawley) [reply benchmark/stats/curve.go]: Nevermind, we already have exceptions to the rule.  Just add a `\|benchmark` to the `grep -v` in `vet.sh` and you should be fine.

Comment (adtac) [reply benchmark/stats/curve.go]: Done.

File: Documentation/benchmark.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+compression turned on, invoking `benchmain` in the following way may closely
+simulate your application:
+
+    go run google.golang.org/grpc/benchmark/benchmain/main.go -workloads=streaming -reqSizeBytes=1024 -respSizeBytes=1024 -compression=gzip
+
+Pass the `-h` flag to the `benchmain` utility to see other flags and workloads that are supported.
+
+## Varying Payload Sizes (Weighted Random Distribution)
+
+The `benchmain` utility supports a `-payloadCurveFile` option that can be used
```
Comment (adtac) [inline Documentation/benchmark.md]: Should update this to `-payloadCurveFiles`

File: Documentation/benchmark.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+(third column). For example, consider the below file:
+
+```csv
+1,32,25
+128,256,25
+1024,2048,50
+```
+
+This tells the `benchmain` utility to generate random RPC requests with a 25%
+probability of payload sizes in the ranges 1-32 bytes, 25% probability in the
```
Comment (adtac) [inline Documentation/benchmark.md]: Should mention that request size = response size.

File: benchmark/benchmain/main.go [hunk 4]
```diff
@@ -86,8 +86,9 @@ var (
 	readKbps           = flags.IntSlice("kbps", defaultReadKbps, "Simulated network throughput (in kbps) - may be a comma-separated list")
 	readMTU            = flags.IntSlice("mtu", defaultReadMTU, "Simulated network MTU (Maximum Transmission Unit) - may be a comma-separated list")
 	maxConcurrentCalls = flags.IntSlice("maxConcurrentCalls", defaultMaxConcurrentCalls, "Number of concurrent RPCs during benchmarks")
-	readReqSizeBytes   = flags.IntSlice("reqSizeBytes", defaultReqSizeBytes, "Request size in bytes - may be a comma-separated list")
-	readRespSizeBytes  = flags.IntSlice("respSizeBytes", defaultRespSizeBytes, "Response size in bytes - may be a comma-separated list")
+	readReqSizeBytes   = flags.IntSlice("reqSizeBytes", nil, "Request size in bytes - may be a comma-separated list")
+	readRespSizeBytes  = flags.IntSlice("respSizeBytes", nil, "Response size in bytes - may be a comma-separated list")
+	payloadCurveFiles  = flags.StringSlice("payloadCurveFiles", nil, "comma-separated list of CSV files describing the shape a random distribution of payload sizes")
```
Comment (adtac) [inline benchmark/benchmain/main.go]: Separate requestPayloadCurveFiles and responsePayloadCurveFiles

**Comment by adtac** [comment]
@dfawley go fmt'd it. Should be green now (is green locally).

File: benchmark/flags/flags.go [hunk 5]
```diff
@@ -138,3 +139,41 @@ func (isv *intSliceValue) String() string {
 	}
 	return b.String()
 }
+
+type stringSliceValue []string
+
+// StringSlice returns a flag representing a slice of strings. Commas will be
+// interpreted as separators unless they are escaped by a backslash.
```
Comment (dfawley) [inline benchmark/flags/flags.go]: This says "backslash" but the comment below says "double-quotes" -- is this right?

File: benchmark/stats/curve.go [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+}
+
+// newPayloadCurveRange receives a line from a payload curve CSV file and
+// returns a *payloadCurveRange if the values are acceptable.
+func newPayloadCurveRange(line []string) (*payloadCurveRange, error) {
+	if len(line) != 3 {
+		return nil, fmt.Errorf("invalid number of entries in line %v (expected 3)", line)
+	}
+
+	var from, to int64
```
Comment (dfawley) [inline benchmark/stats/curve.go]: Should these be int32?  Wouldn't ParseInt fail if it's out of range?

File: benchmark/benchmain/main.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (adtac) [reply benchmark/benchmain/main.go]: Done.

File: benchmark/flags/flags.go (hunk 5 above)
Comment (adtac) [reply benchmark/flags/flags.go]: Done. Removed the comment altogether; the double quotes one is the correct one.

File: benchmark/stats/curve.go (hunk 6 above)
Comment (adtac) [reply benchmark/stats/curve.go]: ParseInt returns a i64, so setting these to i32 would make go complain at compile time.

Comment (dfawley) [reply benchmark/stats/curve.go]: Aha, right.

Actually, given that you're only parsing 32 bits, then checking if it's > `math.MaxInt32` is not necessary.  And I believe `err` will be non-nil when this happens due to out-of-range.

**Pull request merged**